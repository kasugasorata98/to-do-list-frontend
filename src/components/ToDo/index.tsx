import { DeleteIcon, EditIcon } from "@chakra-ui/icons";
import {
  HStack,
  Checkbox,
  Menu,
  MenuButton,
  MenuItem,
  MenuList,
} from "@chakra-ui/react";
import { CSSProperties, Dispatch, SetStateAction, useState } from "react";
import Text from "@/components/Text";
import { ToDoItem } from "@/entities/to-do-list.entity";
import { ApiService } from "@/api";
import { BiDotsHorizontalRounded } from "react-icons/bi";
import Icon from "../Icon";
import Modal from "../Modal";

const ToDo: React.FC<{
  style?: CSSProperties | undefined;
  item: ToDoItem;
  setList: Dispatch<SetStateAction<ToDoItem[]>>;
}> = ({ item, setList }) => {
  const [editOverlay, setEditOverlay] = useState<boolean>(false);

  const updateList = (title: string, isDone: boolean, toDoListId: string) => {
    ApiService.updateList(title, isDone, toDoListId)
      .then(({ data }) => {
        console.log(data);
        setList((prevList) => {
          const newList = [...prevList];
          const index = newList.findIndex((list) => list._id === toDoListId);
          newList[index].isDone = isDone;
          newList[index].title = title;
          return newList;
        });
      })
      .catch((err) => {
        console.log(err);
      })
      .finally(() => {
        setEditOverlay(false);
      });
  };

  const deleteOne = (_id: string) => {
    ApiService.deleteList("DELETE_ONE", _id)
      .then(({ data }) => {
        console.log(data);
        setList((prevList) => {
          const newList = [...prevList];
          return newList.filter((list) => list._id !== _id);
        });
      })
      .catch((err) => {
        console.log(err);
      });
  };

  return (
    <HStack key={item._id} width={"100%"}>
      <Checkbox
        onChange={(e) => {
          updateList(item.title, e.target.checked, item._id);
        }}
        isChecked={item.isDone}
        style={{
          alignSelf: "start",
          marginTop: 4,
          marginLeft: 2,
        }}
      />
      <Text
        style={{
          width: "100%",
          marginTop: -2,
        }}
      >
        {item.title}
      </Text>
      <Modal
        title="Task"
        placeholder="Enter task title"
        header="Edit Title"
        value={item.title}
        isOpen={editOverlay}
        onClose={() => {
          setEditOverlay(false);
        }}
        cancel={{
          title: "Cancel",
          onClick: () => {},
        }}
        confirm={{
          title: "Save",
          onClick: (text) => {
            updateList(text, item.isDone, item._id);
          },
        }}
      />
      <Menu>
        <MenuButton>
          <Icon as={BiDotsHorizontalRounded} />
        </MenuButton>
        <MenuList>
          <MenuItem
            onClick={() => {
              setEditOverlay(true);
            }}
            icon={<EditIcon />}
          >
            Edit
          </MenuItem>
          <MenuItem onClick={() => deleteOne(item._id)} icon={<DeleteIcon />}>
            Delete
          </MenuItem>
        </MenuList>
      </Menu>
    </HStack>
  );
};

export default ToDo;
