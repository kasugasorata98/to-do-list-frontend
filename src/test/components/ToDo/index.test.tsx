import React from "react";
import { render, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import ToDo from "@/components/ToDo";

describe("ToDo component", () => {
  const item = {
    _id: "1",
    title: "Test task",
    isDone: false,
  };
  const setList = jest.fn();

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly", () => {
    const setList = jest.fn();
    const { container } = render(<ToDo item={item} setList={setList} />);
    expect(container).toMatchSnapshot();
  });

  it("should render the todo item correctly", () => {
    const { getByText, getByRole } = render(
      <ToDo item={item} setList={setList} />
    );

    expect(getByText("Test task")).toBeInTheDocument();
    expect(getByRole("checkbox")).not.toBeChecked();
  });

  it("should show the edit overlay when the Edit menu item is clicked", () => {
    const { getByRole, getByText } = render(
      <ToDo item={item} setList={setList} />
    );

    fireEvent.click(getByRole("button"));

    const editMenuItem = getByText("Edit");
    fireEvent.click(editMenuItem);

    expect(getByRole("dialog")).toBeInTheDocument();
  });
});
