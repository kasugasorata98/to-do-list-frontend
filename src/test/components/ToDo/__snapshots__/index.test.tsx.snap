// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDo component renders correctly 1`] = `
<div>
  <div
    class="chakra-stack css-1twuvi0"
  >
    <label
      class="chakra-checkbox css-1elxtiq"
      style="align-self: start; margin-top: 4px; margin-left: 2px;"
    >
      <input
        class="chakra-checkbox__input"
        style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; width: 1px; margin: -1px; padding: 0px; overflow: hidden; white-space: nowrap; position: absolute;"
        type="checkbox"
        value=""
      />
      <span
        aria-hidden="true"
        class="chakra-checkbox__control css-dnty2r"
      />
    </label>
    <p
      class="chakra-text css-0"
      style="color: rgb(226, 234, 241); width: 100%; margin-top: -2px;"
    >
      Test task
    </p>
    <button
      aria-controls="menu-list-:r2:"
      aria-expanded="false"
      aria-haspopup="menu"
      class="chakra-menu__menu-button css-1eqi7qs"
      id="menu-button-:r2:"
    >
      <span
        class="css-xl71ch"
      >
        <svg
          class="chakra-icon css-1dvs3fg"
          fill="currentColor"
          focusable="false"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 10c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm6 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zM6 10c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
          />
        </svg>
      </span>
    </button>
    <div
      class="css-0"
      style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
    >
      <div
        aria-orientation="vertical"
        class="chakra-menu__menu-list css-sjt5nk"
        id="menu-list-:r2:"
        role="menu"
        style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.8) translateZ(0);"
        tabindex="-1"
      >
        <button
          class="chakra-menu__menuitem css-1k756t9"
          data-index="0"
          id="menu-list-:r2:-menuitem-:r3:"
          role="menuitem"
          tabindex="-1"
          type="button"
        >
          <span
            class="chakra-menu__icon-wrapper css-15v20fh"
          >
            <svg
              aria-hidden="true"
              class="chakra-icon chakra-menu__icon css-onkibi"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <g
                fill="none"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-width="2"
              >
                <path
                  d="M11 4H4a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7"
                />
                <path
                  d="M18.5 2.5a2.121 2.121 0 0 1 3 3L12 15l-4 1 1-4 9.5-9.5z"
                />
              </g>
            </svg>
          </span>
          <span
            style="pointer-events: none; flex: 1;"
          >
            Edit
          </span>
        </button>
        <button
          class="chakra-menu__menuitem css-1k756t9"
          data-index="1"
          id="menu-list-:r2:-menuitem-:r4:"
          role="menuitem"
          tabindex="-1"
          type="button"
        >
          <span
            class="chakra-menu__icon-wrapper css-15v20fh"
          >
            <svg
              aria-hidden="true"
              class="chakra-icon chakra-menu__icon css-onkibi"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <g
                fill="currentColor"
              >
                <path
                  d="M19.452 7.5H4.547a.5.5 0 00-.5.545l1.287 14.136A2 2 0 007.326 24h9.347a2 2 0 001.992-1.819L19.95 8.045a.5.5 0 00-.129-.382.5.5 0 00-.369-.163zm-9.2 13a.75.75 0 01-1.5 0v-9a.75.75 0 011.5 0zm5 0a.75.75 0 01-1.5 0v-9a.75.75 0 011.5 0zM22 4h-4.75a.25.25 0 01-.25-.25V2.5A2.5 2.5 0 0014.5 0h-5A2.5 2.5 0 007 2.5v1.25a.25.25 0 01-.25.25H2a1 1 0 000 2h20a1 1 0 000-2zM9 3.75V2.5a.5.5 0 01.5-.5h5a.5.5 0 01.5.5v1.25a.25.25 0 01-.25.25h-5.5A.25.25 0 019 3.75z"
                />
              </g>
            </svg>
          </span>
          <span
            style="pointer-events: none; flex: 1;"
          >
            Delete
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
